Allow using any window as wxStaticBox label

vadz opened the pull request.

vadz committed:
Inline wxStaticBox ctors in wxGTK

No real changes, just make the trivial ctors of this class inline for
consistency with the new ctor about to be added.

vadz committed:
Add support for using arbitrary windows as wxStaticBox labels

This commit implements the new feature in wxGTK and updates the sample
and the documentation.

vadz committed:
Add WXDestroyWithoutChildren() and use it from wxStaticBoxSizer

Factor out the code from wxStaticBoxSizer dtor into a wxStaticBox method
to improve encapsulation: the static box knows better than another class
how to detach its children from it before destroying it.

No real changes yet.

vadz committed:
Add GTKDoApplyWidgetStyle() helper

This allows to call the protected wxWindowGTK::DoApplyWidgetStyle()
method when it's really necessary, e.g. when forwarding to it from
DoApplyWidgetStyle() implementation for another window.

vadz committed:
Move m_labelWin to wxStaticBoxBase itself

It will be reused by all platforms and is not specific to wxGTK.

This also means WXDestroyWithoutChildren() doesn't need to be virtual
any longer.

vadz committed:
Don't make wxStaticBoxBase::GetBordersForSizer() inline

Move the function definition to the source file, there doesn't seem to
be any reason to keep it in the header.

vadz committed:
Remove borders from wxStaticBoxSizer items in the widgets sample

This allows to see better whether the borders returned by
wxStaticBox::GetBordersForSizer() are correct.

vadz committed:
Account for label window in wxStaticBoxBase::GetBordersForSizer()

Take the window used as label into account in the default implementation
of this method.

vadz committed:
Use symbolic names in wxMSW wxStaticBox drawing code

Introduce symbolic constants instead of using raw magic numbers.

No real changes and these numbers are still as magic as before, but at
least they will be easier to change now.

vadz committed:
Slightly decrease the top margin of wxMSW wxStaticBox

Use the same offset we already use in the drawing code, this is a bit
less arbitrary than what we did before and looks slightly better too.

vadz committed:
Adjust wxMSW wxStaticBox pixel constants for DPI

Use FromDIP() with all constants expressed in pixels, this should result
in better appearance when using high DPI.

vadz committed:
Keep attributes after recreating static controls in widgets sample

Preserve the colours, font etc after recreating the widgets to
facilitate testing.

vadz committed:
Don't change wxCompositeWindow size from SetLayoutDirection()

Remove wxSIZE_AUTO from the SetSize() call, this was completely
unnecessary and unexpectedly (and wrongly) resized composite windows
managed by sizers as SetLayoutDirection() side-effect.

vadz committed:
Make wxCompositeWindow ctor protected

As this class is only supposed to be used as a base class, its ctor
doesn't need to be, and hence ought not to be, public.

Also update an outdated comment stating that the ctor didn't do anything
when it, in fact, does perform an important task.

vadz committed:
Document alignment styles in wxStaticBox

Mention that they work in wxGTK, even if it seems to be the only
platform supporting them currently.

vadz committed:
Fix background of wxCheckBoxes inside wxStaticBox in wxMSW

Erase background of the partially transparent native child controls,
such as wxCheckBox, using our own background colour if we have it
instead of using the parent's colour.

For some reason, we -- seemingly intentionally, judging from the comment
-- didn't do it before, but this meant that checkboxes inside static
boxes didn't inherit the box background colour, if it was set, which was
ugly and inconsistent with at least wxGTK. So do use our own background
if we have it now by just reusing the existing PaintBackground() instead
of manually using the parent background brush in WM_PRINTCLIENT handler.

vadz committed:
Add support for arbitrary window labels in wxStaticBox to wxMSW

Just reparent the label window and position it accordingly and, also,
avoid painting over it in MSW-specific code.

vadz committed:
Factor out wxCompositeWindowSettersOnly class

Extract this class from wxCompositeWindow, as sometimes it can be
convenient to just define the setter functions to do the right thing for
a window containing sub-windows, but without dealing with focus too.

This will be used in wxMSW wxStaticBox in the next commit.

vadz committed:
Fix colours and fonts of wxStaticBox label window in wxMSW

Inherit from wxCompositeWindowSettersOnly<> to make sure all the usual
setters, such as SetForegroundColour() and SetFont(), called on
wxStaticBox are propagated to the label window too.

However also prevent SetBackgroundColour() from being propagated
unnecessarily -- because the checkbox already inherits the parent
background colour by default in wxMSW anyhow -- and still override
SetFont() to adjust the label window position after the font change,
otherwise it could be truncated after increasing the font size, for
example.

Because of these issues, wxCompositeWindowSettersOnly is not ideally
suited for its use here, but on balance it still seems to be better to
use it rather than reimplement parts of its functionality here.

vadz committed:
Add support for wxStaticBoxSizer "windowlabel" property to XRC

Allow specifying arbitrary windows as labels for the static boxes
created from XRC.

Note that wxStaticBox XRC handler itself wasn't updated to support this,
as this handler seems to be quite useless because it's impossible to
define any box children with it, so only wxStaticBoxSizer XRC handler
really matters, anyhow.

vadz committed:
Don't disable wxStaticBox window label when disabling the box

This behaviour might be not completely intuitive, but it makes it much
simpler to handle the box state using a checkbox as the label control
(which is by far the most common case of using box window labels).

Notice that while we could add a separate EnableWithoutLabel() method to
wxStaticBox to make it possible to set the state of the box directly
relatively easily, it wouldn't help with using wxEVT_UPDATE_UI for
managing the box state indirectly as it relies on calling Enable() only.
And this solution does allow wxEVT_UPDATE_UI handlers for the box itself
to work (provided the handler takes care to check for the event object
being the box itself, as otherwise it would still disable the child
checkbox when its wxEVT_UPDATE_UI bubbles up to the box).

MaartenBent commented on interface/wx/statbox.h:
```diff
@@ -97,5 +137,51 @@ class wxStaticBox : public wxControl
+
+        @since 3.1.1
+     */
+    wxStaticBox(wxWindow* parent, wxWindowID id,
```
I guess this should be `Create()` instead of the constructor?
And its also available for `wxMSW`.

vadz replied:
Fixed, thanks!

MaartenBent commented:
Seems to build and run fine on Windows with msvc and mingw. Also with a `wxChoice` as label.

It doesn't seem to underline accelerators anymore? E.g. in the widgets sample on the `Spin` page it used to underline the S of `Set style` and the T of `Text alignment`.

vadz committed:
Correct new wxStaticBox::Create() overload documentation

Fix the method name and update its description to mention that it's also
available in wxMSW.

vadz committed:
Move wxHAS_NATIVE_ENABLED_MANAGEMENT definition to the source file

There doesn't seem to be any need to have this symbol in the header,
when it's only used in NotifyWindowOnEnableChange() in wincmn.cpp.

No real changes.

vadz committed:
Always call DoEnable() from NotifyWindowOnEnableChange()

Simplify the code by replacing 2 conditionally-compiled DoEnable() calls
with a single unconditional one.

This doesn't change the behaviour of Enable(), as it always called
DoEnable() on the window itself and only did it for its children when
wxHAS_NATIVE_ENABLED_MANAGEMENT was not defined before and continues to
do the same thing now, but it should fix a small bug in Reparent() as it
didn't update the actual status of the window if it changed as the
result of reparenting before, even though it was supposed to.

vadz committed:
Don't disable wxStaticBox children at wx level when disabling it

Calling Enable() on all children from wxStaticBox::Enable() was wrong,
the actual status of the child, returned by wxWindow::IsThisEnabled(),
is not supposed to change just because its parent was disabled.

Call NotifyWindowOnEnableChange() to avoid this, while still disabling
the children visually.

vadz committed:
Fix bug with not re-enabling wxStaticBox with label sometimes

In some specific scenario, described in the newly added comment in
wxStaticBoxBase::Enable(), the box and its label could remain disabled
after its parent was disabled and re-enabled.

Fix this by continuing to use the derived class version for disabling
the box children, but not when enabling them, as the base class version
already does the right thing in this case.

vadz commented:
Thanks for reviewing and testing! I don't know what has happened with the labels, they shouldn't have been affected by these changes.

I did find another nasty bug with the box enabled state in the meanwhile, which I fixed now (hopefully without breaking anything else).

One related remaining problem is that, as I've noticed only recently, MSW doesn't actually grey out the labels of disabled group boxes. I wonder if we should do it manually because they just don't look right otherwise.

vadz committed:
Don't pretend static box with enabled label is disabled

Trying to be smart by setting m_isEnabled to false in
wxStaticBox::Enable() without actually disabling the box itself (because
it can't be done if its label window is to remain enabled) didn't really
work. For example, it was impossible to TAB to a checkbox label of the
box when it was disabled, because keyboard navigation (correctly)
doesn't recurse into disabled windows and there could be similar
problems with any other code iterating over windows and skipping over
the disabled ones.

So, finally, simplify things and keep m_isEnabled in sync with the real
box state, even if this, counter-intuitively, means that IsEnabled() on
the box returns true after calling Enable(false) on it.

This also reverts 4ee35cf5ee569b6ee6c7d0d5702484d4d2a20f96 ("Don't
disable wxStaticBox children at wx level when disabling it") as we can't
avoid really disabling the children any more now that their parent is
not disabled: without this, their IsEnabled() would return true, i.e.
they wouldn't be disabled at all, from the program point of view. This
is unfortunate for the reasons that originally motivated that commit,
i.e. if some wxStaticBox child is disabled, disabling and re-enabling
the box will now re-enable this child, even if it shouldn't, but seems
impossible to avoid. The only possible alternative is to modify
IsEnabled() to add some wxStaticBox-specific hook to it, e.g. instead of
calling GetParent()->IsEnabled() there, we could call some now
AreChildrenEnable() method, which would delegate to IsEnabled() by
default but overridden in wxStaticBox. However this seems complicated,
and will add an extra virtual function call to all (frequently
happening) IsEnabled() calls.
